### PR TITLE
Add joinability for groups

### PR DESCRIPTION
**Pull request by dbkr**
Adds API to set the 'joinable' flag, and corresponding flag in the
table.

### CODE REVIEW

File: synapse/groups/groups_server.py [hunk 1]
```diff
@@ -205,6 +206,24 @@ def delete_group_summary_room(self, group_id, requester_user_id,
...
+    @defer.inlineCallbacks
+    def set_group_joinable(self, group_id, requester_user_id, content):
+        """Sets whether a group is joinable without an invite or knock
+        """
+        yield self.check_group_is_ours(
+            group_id, requester_user_id, and_exists=True, and_is_admin=requester_user_id
+        )
+
+        is_joinable = content.get('joinable')
+        if is_joinable is None:
```
Comment (krombel) [inline synapse/groups/groups_server.py]: Shouldn't we also check if is_joinable evaluates to a boolean?
Or is it enough to handle that on a database level as storing arbitrary data into a boolean field will fail?

File: synapse/rest/client/v2_alpha/groups.py [hunk 2]
```diff
@@ -401,6 +402,32 @@ def on_GET(self, request, group_id):
...
+        super(GroupJoinableServlet, self).__init__()
+        self.auth = hs.get_auth()
+        self.groups_handler = hs.get_groups_local_handler()
+
+    @defer.inlineCallbacks
+    def on_POST(self, request, group_id):
+        requester = yield self.auth.get_user_by_req(request)
+        requester_user_id = requester.user.to_string()
+
+        content = parse_json_object_from_request(request)
```
Comment (krombel) [inline synapse/rest/client/v2_alpha/groups.py]: shouldn't we evaluate that we have only have `{"joinable": <bool>}` as body?

**Comment by dbkr** [comment]
I assume this is a flaky sytest?

**Comment by krombel** [comment]
@dbkr I saw that specific test failing on other PR's as well => I think yes

File: synapse/federation/transport/server.py [hunk 3]
```diff
@@ -1124,6 +1125,24 @@ def on_POST(self, origin, content, query):
         defer.returnValue((200, resp))
 
 
+class FederationGroupsJoinableServlet(BaseFederationServlet):
+    """Sets whether a group is joinable without an invite or knock
+    """
+    PATH = "/groups/(?P<group_id>[^/]*)/joinable$"
+
+    @defer.inlineCallbacks
+    def on_POST(self, origin, content, query, group_id):
```
Comment (dbkr) [inline synapse/federation/transport/server.py]: Should this be PUT or POST? The client API is a PUT.

Comment (lukebarnard1) [reply synapse/federation/transport/server.py]: Erik said it should have the same method as the CS API. @erikjohnston, can you confirm?

**Comment by lukebarnard1** [comment]
https://github.com/matrix-org/sytest/pull/436 will need updating once these changes have landed.

File: synapse/storage/group_server.py [hunk 4]
```diff
@@ -29,6 +30,18 @@
 
 
 class GroupServerStore(SQLBaseStore):
+    def set_group_join_policy(self, group_id, join_policy):
+        return self._simple_update_one(
```
Comment (erikjohnston) [inline synapse/storage/group_server.py]: `set_group_join_policy` needs a docstring, especially as we only want to accept certain values so we should document what they are.

Might also be worth adding a `assert join_policy in ('invite', 'open')` (we should probably move some of those to constants somewhere at some point)

File: synapse/storage/schema/delta/48/groups_joinable.sql [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ * limitations under the License.
+ */
+
+/* 
+ * This isn't a real ENUM because sqlite doesn't support it
+ * and we use a default of NULL for inserted rows and interpret
+ * NULL at the python store level as necessary so that existing
+ * rows are given the correct default policy.
+ */
+ALTER TABLE groups ADD COLUMN join_policy TEXT DEFAULT NULL;
```
Comment (erikjohnston) [inline synapse/storage/schema/delta/48/groups_joinable.sql]: If we can have defaults then we may as well do

```sql
ALTER TABLE groups ADD COLUMN join_policy TEXT NON NULL DEFAULT 'invite';
```

File: synapse/federation/transport/client.py [hunk 6]
```diff
@@ -856,6 +857,21 @@ def update_group_summary_user(self, destination, group_id, requester_user_id,
             ignore_backoff=True,
         )
 
+    @log_function
+    def set_group_joinable(self, destination, group_id, requester_user_id,
+                           content):
+        """Sets the join policy for a group
+        """
+        path = PREFIX + "/groups/%s/setting/m.join_policy" % (group_id,)
```
Comment (erikjohnston) [inline synapse/federation/transport/client.py]: (I was expecting it to be `settings` plural, but that's a matter of taste)

**Comment by erikjohnston** [comment]
(I didn't mean to assign @richvdh there....)

**Pull request merged**